### ISSUE TITLE
resnet example performance issues

### ISSUE THREAD

**Issue by bhetherman** [flexflow/FlexFlow#62]
When running the resnet cpp example on a single V100 GPU with batch_size of 16 (I assume its fp32 as well since FlexFlow doesnt seem to have an fp16 option as far as I know) I'm seeing throughput of 30-32 samples per second. This means an iteration time of about 500ms.

In pytorch I get an iteration time for fp32, batch 16 and same input size of 58ms. Thats a pretty large ~10x discrepancy. Is an iteration time of 500 ms expected? Is there any options to FlexFlow i can add to increase performance?

Looking at the individual layer times from measure_compute_time seem closer to pytorch. So I dont understand why the actual run time is so much slower. Especially when you take parallelization out of the equation.

**Comment by jiazhihao** [comment]
@bhetherman One thing it check first is whether you built FlexFlow in debug or release mode (debug is the default). Can you verify that `DEBUG` is set to 0 in the Makefile?

**Comment by bhetherman** [comment]
@jiazhihao Good call, i thought it was 0 but looks like it was 1. Ill get back to you with new numbers soon

**Comment by bhetherman** [comment]
@jiazhihao just rebuilt and ran the network. That was definitely a huge part of the issue. Now Im seeing throughput of about 93-94 samples/s. So about 170ms per iteration, so not 10x slower but still about 3x.

Any other tips or tricks?

Ps I just using synthetic data right now

**Comment by jiazhihao** [comment]
@bhetherman Sorry that I didn't get back to you on this last week. Which ResNet example did you used (C++ or Python one)? There shouldn't be observable performance difference between FlexFlow and PyTorch on a single GPU.

**Comment by bhetherman** [comment]
@jiazhihao no worries. Im using the C++ resnet example on a single V100

**Comment by bhetherman** [comment]
@jiazhihao still having performance issues, can you share what kind of throughput/iteration times you are seeing for the cpp resnet example? Thanks.

**Comment by jiazhihao** [comment]
@bhetherman I tried this on a V100 GPU (with CUDA 9.2 and cuDNN 7.5). With a batch size of 16, the training throughput is 210 images/second. The per-iteration execution time is 76ms. The simulator predicts a per-iteration runtime of around 50ms, which means there is a non-negligible runtime overhead for running ResNet. One potential cause is that when per-operator execution time is small (e.g., using a small batch size), the DNN computation cannot effectively hide all Legion dynamic analysis overhead, making FlexFlow unable to achieve the peek performance for small batch runs.

I am generating a few Legion prof logs to verify this.

**Comment by bhetherman** [comment]
@jiazhihao Ok I think I identified the issue. It seams the batch norm layer causes a large performance drop on its own. But I was also separating out the relu into another layer which caused an even further performance drop.

For 1 V100, batch 16:
Just Conv layer: 214.38 samples/s
Conv layer + batch norm layer: 140 samples/s
Conv layer + batch norm layer + Relu layer: 120 samples/s

Do you think its FlexFlow or Legion overhead that causes these ops to affect performance so much? Because they really shouldn't take that much time i would think.

Thanks.

**Comment by ghostplant** [comment]
@bhetherman Have you tried AlexNet with multiple nodes? Is it faster than single node?
